superduperdb/ext/llm/model.py

```diff
@@ -0,0 +1,440 @@
+from transformers import (
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    BitsAndBytesConfig,
+    TrainerCallback,
+    TrainingArguments,
+)
+
+from superduperdb import logging
+from superduperdb.backends.query_dataset import query_dataset_factory
+from superduperdb.components.dataset import Dataset as _Dataset
+from superduperdb.components.model import (
+    Model,
+    _TrainingConfiguration,
+)
+from superduperdb.ext.utils import ensure_initialized
+
+if typing.TYPE_CHECKING:
+    from superduperdb.backends.base.query import Select
+    from superduperdb.base.datalayer import Datalayer
+    from superduperdb.components.metric import Metric
+
+
+class LLMCallback(TrainerCallback):
+    def __init__(self, llm: "LLM"):
+        self.llm = llm
+
+    def on_evaluate(self, args, state, control, **kwargs):
+        self.llm.append_metrics(state.log_history[-1])
+
+
+@dc.dataclass
+class LLMTrainingArguments(TrainingArguments):
+    """
+    LLM Training Arguments.
+    Inherits from :class:`transformers.TrainingArguments`.
+
+    {training_arguments_doc}
+        lora_r (`int`, *optional*, defaults to 8):
+            Lora R dimension.
+
+        lora_alpha (`int`, *optional*, defaults to 16):
+            Lora alpha.
+
+        lora_dropout (`float`, *optional*, defaults to 0.05):
+            Lora dropout.
+
+        lora_target_modules (`List[str]`, *optional*, defaults to None):
+            Lora target modules. If None, will be automatically inferred.
+
+        lora_weight_path (`str`, *optional*, defaults to ""):
+            Lora weight path.
+
+        lora_bias (`str`, *optional*, defaults to "none"):
+            Lora bias.
+
+        max_length (`int`, *optional*, defaults to 512):
+            Maximum source sequence length during training.
+
+    """
+
+    lora_r: int = 8
+    lora_alpha: int = 16
+    lora_dropout: float = 0.05
+    lora_target_modules: t.Optional[t.List[str]] = None
+    lora_weight_path: str = ""
+    lora_bias: str = "none"
+    max_length: t.Optional[int] = 512
+
+    __doc__ = __doc__.format(training_arguments_doc=TrainingArguments.__doc__)
+
+
+@functools.wraps(LLMTrainingArguments)
+def LLMTrainingConfiguration(identifier: str, **kwargs) -> _TrainingConfiguration:
+    return _TrainingConfiguration(identifier=identifier, kwargs=kwargs)
+
+
+@dc.dataclass
+class LLM(Model):
+    """
+    LLM model based on `transformers` library.
+    Parameters:
+    : param identifier: model identifier
+    : param model_name_or_path: model name or path
+    : param bits: quantization bits, [4, 8], default is None
+    : param model_kwargs: model kwargs,
+        all the kwargs will pass to `transformers.AutoModelForCausalLM.from_pretrained`
+    : param tokenizer_kwags: tokenizer kwargs,
+        all the kwargs will pass to `transformers.AutoTokenizer.from_pretrained`
+    """
+
+    identifier: str = ""
+    model_name_or_path: str = "facebook/opt-125m"
+    bits: t.Optional[int] = None
```

Comment:
Do we assume that user has quantised the model to 8/4 bits?

Reply:
No, not quantized by default, keeping the default initialization behavior of transformers